src/rosout.rs

```diff
@@ -0,0 +1,124 @@
+use ringbuf;
+use rosrust;
+use std::option::Option;
+use std::sync::{Arc, RwLock};
+
+pub struct LogBuffer {
+    producer: ringbuf::Producer<String>,
+    consumer: ringbuf::Consumer<String>,
+    is_buffering: bool,
+}
+
+impl LogBuffer {
+    pub fn new(size: usize, start_buffering: bool) -> LogBuffer {
+        let (producer, consumer) = ringbuf::RingBuffer::<String>::new(size).split();
+        LogBuffer {
+            producer,
+            consumer,
+            is_buffering: start_buffering,
+        }
+    }
+
+    pub fn push_logstring(&mut self, logstring: String) {
+        if !self.is_buffering {
+            return;
+        }
+        if self.producer.remaining() == 0 {
+            self.consumer.pop();
+        }
+        if self.producer.push(logstring).is_err() {
+            eprintln!("failed to buffer rosout log")
+        }
+    }
+
+    pub fn read_logstrings(&self) -> std::vec::Vec<String> {
+        let mut logstrings = std::vec::Vec::<String>::new();
+        self.consumer.for_each(|logstring| {
+            logstrings.push(logstring.clone());
+        });
+        return logstrings;
+    }
+}
+
+fn format_logstring(
+    level: i8,
+    msg: &str,
+    name: &str,
+    stamp: &rosrust::Time,
+    min_loglevel: i8,
+) -> Option<String> {
+    use colored::{Color, Colorize};
+    use rosrust_msg::rosgraph_msgs::Log;
+
+    if level < min_loglevel {
+        return None;
+    }
+
+    let format_string = |levelname, color| {
```

Comment:
fancy

Reply:
using 1.58 formatting for more fancyness